distributed/shuffle/_disk.py

```diff
@@ -96,6 +107,94 @@ def read(self, id: int | str) -> bytes:
         else:
             raise KeyError(id)
 
+
+class _PersistentBytesIO(BytesIO):
+    """A BytesIO object that does not close itself when used in a with block."""
+
+    def __enter__(self) -> _PersistentBytesIO:
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        pass
+
+
+class MemoryShardsBuffer(FileShardsBuffer):
+    """Accept and buffer many small objects into memory.
+
+    This implements in-memory "file" buffering with no resource limit
+    with the same interface as :class:`DiskShardsBuffer`.
+
+    """
+
+    def __init__(self) -> None:
+        super().__init__(memory_limiter=None)
+        self._memory_buffers: defaultdict[str, _PersistentBytesIO] = defaultdict(
+            _PersistentBytesIO
+        )
+
+    def writer(self, id: int | str) -> BinaryIO:
+        buf = self._memory_buffers[str(id)]
+        buf.seek(buf.tell())
+        return buf
+
+    def reader(self, id: int | str) -> BinaryIO:
+        key = str(id)
+        if key not in self._memory_buffers:
+            raise FileNotFoundError(f"Shard with {id=} is unknown")
+        buf = self._memory_buffers[str(id)]
```

Comment:
```suggestion
        buf = self._memory_buffers.pop(str(id))
```

This will remove the `BinaryIO` object from the `MemoryShardsBuffer` upon reading to prevent memory duplication after unpacking. From what I understand, reading should be an exactly-once operation.